### PR TITLE
Add basic text tokenized

### PR DESCRIPTION
**Pull request by zechengz**
Add basic text tokenized functionalities. Encoder change to be in the following PR.

### CODE REVIEW

File: test/datasets/test_fake_dataset.py [hunk 1]
```diff
@@ -50,3 +61,9 @@ def test_fake_dataset(with_nan):
     assert feat_text_embedded.shape == (
         num_rows, len(tensor_frame.col_names_dict[torch_frame.text_embedded]),
         out_channels)
+
+    feat_text_tokenized: Dict[str, MultiNestedTensor] = tensor_frame.feat_dict[
```
Comment (yiweny) [inline test/datasets/test_fake_dataset.py:65]:
```suggestion
    feat_text_tokenized: TensorData = tensor_frame.feat_dict[
```

File: torch_frame/config/text_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    Args:
+        text_tokenizer (callable): A callable text tokenizer that takes a list
+            of strings as input and output the dictionary of
+            :class:`MultiNestedTensor`.
+        batch_size (int, optional): Batch size to use when encoding the
+            sentences. If set to :obj:`None`, the text embeddings will
+            be obtained in a full-batch manner. (default: :obj:`None`)
+
+    """
+    text_tokenizer: Callable[[List[str]], Dict[str, MultiNestedTensor]]
```
Comment (yiweny) [inline torch_frame/config/text_tokenizer.py:21]:
```suggestion
    text_tokenizer: Callable[[List[str]], TensorData]
```

File: torch_frame/data/tensor_frame.py [hunk 3]
```diff
@@ -83,6 +83,11 @@ def validate(self):
             num_cols = len(self.col_names_dict[stype_name])
             if num_cols == 0:
                 empty_stypes.append(stype_name)
+
+            # Get any value if feat is a dictionary
+            if isinstance(feat, dict):
+                feat = next(iter(feat.values()))
```
Comment (yiweny) [inline torch_frame/data/tensor_frame.py:89]: It's better to iterate the keys and check for all values.

File: torch_frame/data/dataset.py [hunk 4]
```diff
@@ -184,11 +203,19 @@ def __call__(
...
-        feat_dict = {
-            stype: (torch.stack(xs, dim=1) if not stype.use_multi_nested_tensor
-                    else MultiNestedTensor.stack(xs, dim=1))
-            for stype, xs in xs_dict.items()
-        }
+
+        feat_dict = {}
+        for stype, xs in xs_dict.items():
+            if stype.use_multi_nested_tensor:
+                feat_dict[stype] = MultiNestedTensor.stack(xs, dim=1)
```
Comment (yiweny) [inline torch_frame/data/dataset.py:210]:
```suggestion
                feat_dict[stype] = MultiNestedTensor.cat(xs, dim=1)
```

File: torch_frame/data/mapper.py [hunk 5]
```diff
@@ -209,3 +209,59 @@ def forward(
...
+
+
+class TextTokenizationTensorMapper(TensorMapper):
+    r"""Tokenize any text series into a dictionary of
+    :class:`MultiNestedTensor`.
+
+    Args:
+        text_tokenizer (callable): A callable function that takes list of
+            strings and returns a dictionary of :class:`MultiNestedTensor`.
+            The values of the dictionary contains for example tokens and
```
Comment (yiweny) [inline torch_frame/data/mapper.py:221]: Correct me if I am wrong, but I think the set of keys should be the same for all tokenizer, i.e. ids, attention_masks and tokens. Can you be more specific in the documentation.

File: torch_frame/data/mapper.py [hunk 6]
```diff
@@ -209,3 +209,59 @@ def forward(
...
+            attention masks.
+        batch_size (int, optional): The mini-batch size used for the text
```
Comment (yiweny) [inline torch_frame/data/mapper.py:223]: Just curious, why is mini-batch necessary if we are not having parallelism

File: torch_frame/data/multi_nested_tensor.py [hunk 7]
```diff
@@ -28,7 +28,6 @@ def __init__(
     ):
         assert offset[0] == 0
         assert offset[-1] == len(values)
-        assert len(offset) == num_rows * num_cols + 1
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:31]: Probably good to keep this?

File: torch_frame/data/multi_nested_tensor.py [hunk 8]
```diff
@@ -224,6 +231,29 @@ def stack(xs: List['MultiNestedTensor'],
         else:
             raise NotImplementedError
 
+    @staticmethod
+    def cat(xs: List['MultiNestedTensor'],
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:235]: I think in PyTorch `Tuple` of `Tensors` are also supported. We should implement this function with the same arguments as PyTorch `cat`.

File: torch_frame/data/multi_nested_tensor.py [hunk 9]
```diff
@@ -224,6 +231,29 @@ def stack(xs: List['MultiNestedTensor'],
...
+            num_cols = sum(x.num_cols for x in xs)
+        else:
+            raise NotImplementedError
+        values = torch.cat([x.values for x in xs], dim=0)
+        accum_offset = 0
+        offset = []
+        for x in xs:
+            offset.append(x.offset + accum_offset)
+            accum_offset = accum_offset + x.offset[-1]
+        offset = torch.cat(offset, dim=0)
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:253]: `x.offset` contains a 0 index. It won't work with this logic because a dummy 0 will be added for each `x` in `xs`.

**Review by yiweny — CHANGES_REQUESTED** [review]
I think you need to add a test case for `cat` function in `MultiNestedTensor`. Also there is some issue with the logic.

File: torch_frame/data/multi_nested_tensor.py [hunk 10]
```diff
@@ -224,6 +231,29 @@ def stack(xs: List['MultiNestedTensor'],
...
+            num_rows = xs[0].num_rows
+            num_cols = sum(x.num_cols for x in xs)
+        else:
+            raise NotImplementedError
+        values = torch.cat([x.values for x in xs], dim=0)
+        accum_offset = 0
+        offset = []
+        for x in xs:
+            offset.append(x.offset + accum_offset)
+            accum_offset = accum_offset + x.offset[-1]
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:252]: This logic is only applicable to `dim=0`.

**Comment by weihua916** [comment]
Thanks @zechengz As @yiweny said, it'd be great if you could separate `cat` method out and thoroughly test it in both `dim=0` and `dim=1`.

File: torch_frame/data/mapper.py (hunk 5 above)
Comment (zechengz) [reply torch_frame/data/mapper.py:221]: No, they are not the same. For example BERT has attention mask, token input ids and sequence id. But roberta only has attention mask and token input ids. Will update the docs.

File: torch_frame/data/mapper.py (hunk 6 above)
Comment (zechengz) [reply torch_frame/data/mapper.py:223]: Tokenizer does not have this issue as they don't need GPU. For convenience I just keep the interface the same. @weihua916 Do you think we can remove the `batch_size` for text tokenized?

File: torch_frame/data/multi_nested_tensor.py (hunk 10 above)
Comment (zechengz) [reply torch_frame/data/multi_nested_tensor.py:252]: Right, moved to `dim=0`

File: torch_frame/data/multi_nested_tensor.py (hunk 9 above)
Comment (zechengz) [reply torch_frame/data/multi_nested_tensor.py]: Right, will be fixed in the following PR

File: torch_frame/data/mapper.py (hunk 6 above)
Comment (zechengz) [reply torch_frame/data/mapper.py:224]: Discussed with @weihua916 , for now just keep it same with text embedder cfg. In the future we can easily delete it.

File: torch_frame/data/multi_nested_tensor.py [hunk 11]
```diff
@@ -358,6 +358,23 @@ def stack(xs: List['MultiNestedTensor'],
...
+            return xs[0]
+
+        if dim < 0:
+            dim += xs[0].ndim
+
+        if dim == 0:
+            raise NotImplementedError
+        elif dim == 1:
+            raise NotImplementedError
+        else:
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:375]: Do we need to support cat on the third dimension? Maybe raise a different error here.

File: torch_frame/data/mapper.py [hunk 12]
```diff
@@ -209,3 +209,60 @@ def forward(
...
+        device: Optional[torch.device] = None,
+    ) -> Dict[str, MultiNestedTensor]:
+        ser = ser.astype(str)
+        ser_list = ser.tolist()
+
+        seq_tokens_dict: Optional[Dict[str, MultiNestedTensor]] = None
+        if self.batch_size is None:
+            seq_tokens_dict = self.text_tokenizer(ser_list)
+            for key in seq_tokens_dict:
+                seq_tokens_dict[key].to(device)
```
Comment (weihua916) [inline torch_frame/data/mapper.py:250]:
```python
for key, val in seq_tokens_dict:
    seq_tokens_dict[key] = value.to(device)
```

**Review by yiweny — APPROVED** [review]

File: torch_frame/data/tensor_frame.py [hunk 13]
```diff
@@ -83,20 +83,42 @@ def validate(self):
...
+                            f"feature is {num_cols}, which does not align "
+                            f"with the column dimensionality of "
+                            f"feat_dict['{stype_name}']['{key}'] "
+                            f"(got {feat.size(1)})")
+                    if feat.size(0) != num_rows:
+                        raise ValueError(
+                            f"The length of elements in "
+                            f"feat_dict['{stype_name}']['{key}']"
+                            f"is not aligned, got {feat.size(0)} but "
+                            f"expected {num_rows}.")
```
Comment (weihua916) [inline torch_frame/data/tensor_frame.py:121]: This is highly repetitive... Can we somehow make this cleaner?

File: torch_frame/testing/text_tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+                 device: Optional[torch.device] = None):
+        self.device = device
+        self.num_hash_bins = num_hash_bins
+
+    def __call__(self, sentences: List[str]) -> Dict[str, MultiNestedTensor]:
+        res = []
+        for s in sentences:
+            tokens = s.split(' ')
+            idx = torch.LongTensor(
+                [hash(t) % self.num_hash_bins for t in tokens])
```
Comment (weihua916) [inline torch_frame/testing/text_tokenizer.py:29]: torch.tensor

File: torch_frame/data/mapper.py [hunk 15]
```diff
@@ -209,3 +209,60 @@ def forward(
...
+    def forward(
+        self,
+        ser: Series,
+        *,
+        device: Optional[torch.device] = None,
+    ) -> Dict[str, MultiNestedTensor]:
+        ser = ser.astype(str)
+        ser_list = ser.tolist()
+
+        seq_tokens_dict: Optional[Dict[str, MultiNestedTensor]] = None
```
Comment (weihua916) [inline torch_frame/data/mapper.py:246]: weird to set it to None at the beggining.

File: torch_frame/data/mapper.py [hunk 16]
```diff
@@ -209,3 +209,60 @@ def forward(
...
+            if seq_tokens_dict is None or len(seq_tokens_dict) == 0:
+                seq_tokens_dict = batch_seq_tokens_dict
+            else:
+                for key in seq_tokens_dict:
+                    seq_tokens_dict[key] = MultiNestedTensor.cat(
+                        [seq_tokens_dict[key], batch_seq_tokens_dict[key]],
+                        dim=0)
+        for key in seq_tokens_dict:
+            seq_tokens_dict[key].to(device)
+        return seq_tokens_dict
```
Comment (weihua916) [inline torch_frame/data/mapper.py:265]: re-write so that we do concatenation at last.

File: torch_frame/data/multi_nested_tensor.py (hunk 11 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:375]: We should raise an error.

**Review by weihua916 — COMMENTED** [review]
Thanks, although I found this PR is severely untested.

From next time, we should implement things sequentially and test PR more thoroughly (first implement `cat(dim=1)`, then this functionality).

File: torch_frame/data/mapper.py (hunk 12 above)
Comment (weihua916) [reply torch_frame/data/mapper.py:250]: same for other places.

File: torch_frame/data/tensor_frame.py (hunk 13 above)
Comment (weihua916) [reply torch_frame/data/tensor_frame.py:121]: You can make this as a function block and call it twice.
```python
if feat.dim() < 2:
                        raise ValueError(f"feat_dict['{stype_name}']['{key}'] "
                                         f"must be at least 2-dimensional")
                    if num_cols != feat.size(1):
                        raise ValueError(
                            f"The expected number of columns for {stype_name} "
                            f"feature is {num_cols}, which does not align "
                            f"with the column dimensionality of "
                            f"feat_dict['{stype_name}']['{key}'] "
                            f"(got {feat.size(1)})")
                    if feat.size(0) != num_rows:
                        raise ValueError(
                            f"The length of elements in "
                            f"feat_dict['{stype_name}']['{key}']"
                            f"is not aligned, got {feat.size(0)} but "
                            f"expected {num_rows}.")
```
